File: 207.course-schedule/bfs.rb
#
# @lc app=leetcode.cn id=207 lang=ruby
# @lcpr version=30204
#
# [207] 课程表
#

# @lcpr-template-start

# @lcpr-template-end
# @lc code=start
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Boolean}
def can_finish(num_courses, prerequisites)
  graph = Array.new(num_courses) { [] }
  in_degrees = Array.new(num_courses, 0)
  prerequisites.each do |(u, v)|
    graph[v].push(u)
    in_degrees[u] += 1
  end
  queue = []
  in_degrees.each_with_index do |d, i|
    queue << i if d == 0
  end

  while queue.size > 0
    n = queue[0]
    queue = queue[1..]

    graph[n].each do |nn|
      in_degrees[nn] -= 1
      queue << nn if in_degrees[nn] == 0
    end
  end

  in_degrees.none? { |d| d > 0 }
end

# @lc code=end

#
# @lcpr case=start
# 2\n[[1,0]]\n
# @lcpr case=end

# @lcpr case=start
# 2\n[[1,0],[0,1]]\n
# @lcpr case=end

# @lcpr case=start
# 4\n[[1,0],[1,2],[0,2]]\n
# @lcpr case=end

# @lcpr case=start
# 3\n[[0,1],[0,2],[1,2],[2,1]]\n
# @lcpr case=end

#

File: 207.course-schedule/bfs.js
/*
 * @lc app=leetcode.cn id=207 lang=javascript
 * @lcpr version=30204
 *
 * [207] 课程表
 */


// @lcpr-template-start

// @lcpr-template-end
// @lc code=start
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
  const inAdj = Array.from({ length: numCourses }, () => []);
  const outAdj = Array.from({ length: numCourses }, () => []);
  prerequisites.forEach(([u, v]) => {
    inAdj[u].push(v);
    outAdj[v].push(u);
  })
  const inDegrees = inAdj.map(adj => adj.length);
  const queue = [];
  for (let i = 0; i < numCourses; i++) {
    if (inDegrees[i] === 0) {
      queue.push(i);
    }
  }
  let queue_i = 0;
  while (queue_i < queue.length) {
    const i = queue[queue_i++];
    for (const j of outAdj[i]) {
      inDegrees[j]--;
      if (inDegrees[j] === 0) {
        queue.push(j);
      }
    }
  }

  for (let i = 0; i < numCourses; i++) {
    if (inDegrees[i] !== 0) {
      return false;
    }
  }

  return true;
};
// @lc code=end



// @lcpr case=start
// 2\n[[1,0]]\n
// @lcpr case=end

// @lcpr case=start
// 2\n[[1,0],[0,1]]\n
// @lcpr case=end


